prpc/db.py

```diff
@@ -0,0 +1,74 @@
+# SPDX-License-Identifier: GPL-3.0-or-later
+import sqlite3
+
+
+# This class relies on the __del__ method of the SQLite connection object
+# to properly clean up the connections. It's not handled directly in the class
+# for this reason.
+class DBConnection:
+    def __init__(self, db_path):
+        """
+        Manage the database as a context manager.
+
+        :param str db_path: the path to the database that the app is configured to use
+        """
+        self._db_path = db_path
+        self._connection = None
+
+    @property
+    def connection(self):
+        """Return a started connect to the database."""
+        if not self._connection:
+            self._connection = sqlite3.connect(f"file:{self._db_path}", uri=True)
+        return self._connection
+
+    def create_tables(self):
+        """Create the database tables for the application."""
+        cursor = self.connection.cursor()
+        # Note that SQLite does not support booleans so an integer is used for the success column
+        cursor.execute(
+            """
+            CREATE TABLE IF NOT EXISTS pipelineRuns (
+                id INTEGER PRIMARY KEY,
+                name TEXT NOT NULL,
+                success INTEGER NOT NULL
+            )
+            """
+        )
+        cursor.execute("CREATE INDEX IF NOT EXISTS name_idx ON pipelineRuns (name)")
+        cursor.execute("CREATE INDEX IF NOT EXISTS success_idx ON pipelineRuns (success)")
```

Comment:
I'm unsure how efficient those indexes are - there are only few values for name, and two possible values for success.

How about instead using something like:

```sql
-- success/failure counters
CREATE TABLE IF NOT EXISTS pipelineRuns (
  name TEXT PRIMARY KEY,
  success INTEGER NOT NULL,
  failure INTEGER NOT NULL
);

-- Increment success example:
INSERT INTO pipelineRuns (name, success, failure)
  VALUES('run_prod_updater', 1, 0) 
  ON CONFLICT(name) 
  DO UPDATE SET success = success + 1;

-- Increment failure example:
INSERT INTO pipelineRuns (name, success, failure)
  VALUES('run_prod_updater', 0, 1) 
  ON CONFLICT(name) 
  DO UPDATE SET failure = failure + 1;
```

Reply:
@hluk the indexes help for the group by query since those must be sorted by name and success value into groups.

As for your suggestion, I didn't do something similar since I was worried about multiple writes and causing race conditions. Since SQLite only allows a single writer at a time, I can give it a shot.